File: backend/src/middleware/auth_middleware.py
from fastapi import Request, HTTPException, status
from typing import Dict, Any
import jwt
from ..utils.database import get_session
from sqlmodel import Session


def get_current_user(request: Request) -> Dict[str, Any]:
    """
    Extract and verify user from request headers.

    In a real implementation, this would extract and verify JWT token
    from Authorization header. For demo purposes, we'll simulate authentication.
    """
    # In a real implementation, we would:
    # 1. Extract Authorization header
    # 2. Verify JWT token
    # 3. Return user info

    # For demo purposes, we'll return a mock user
    # In production, you would verify the token with your auth provider
    auth_header = request.headers.get("Authorization")

    if not auth_header or not auth_header.startswith("Bearer "):
        # For demo purposes, allow requests without auth but return mock user
        # In production, this would raise an exception
        return {"id": "demo-user-id", "email": "demo@example.com"}

    # Extract token
    token = auth_header.split(" ")[1]

    # In a real implementation, verify the JWT
    # decoded_token = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
    # return decoded_token["user"]

    # For demo purposes, return mock user
    return {"id": "demo-user-id", "email": "demo@example.com"}


def verify_user_access(user_id: str, requested_user_id: str) -> bool:
    """
    Verify that the authenticated user can access the requested resource.

    Args:
        user_id: ID of the authenticated user
        requested_user_id: ID of the user whose resource is being requested

    Returns:
        Boolean indicating if access is allowed
    """
    # In a real implementation, this would compare the user IDs
    # For demo purposes, we'll allow access to demo user
    return user_id == requested_user_id or user_id == "demo-user-id"

File: src/app/api/middleware/auth.middleware.ts
import { MiddlewareHandler } from 'hono';
import { verify } from 'hono/jwt';

// In a real implementation, this would integrate with Better Auth
// For now, we'll implement a simplified version
export const authMiddleware: MiddlewareHandler = async (c, next) => {
  // In a real implementation, extract and verify JWT token
  // For demo purposes, we'll simulate user authentication

  try {
    // Get token from header
    const authHeader = c.req.header('Authorization');

    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      return c.json({ error: 'Authorization header missing or invalid' }, 401);
    }

    const token = authHeader.substring(7); // Remove 'Bearer ' prefix

    // In a real implementation, verify the JWT token
    // const decoded = await verify(token, Bun.env.JWT_SECRET || '');
    // c.set('userId', decoded.userId);

    // For demo purposes, we'll simulate authentication
    // In a real app, you'd verify the JWT with your auth provider
    c.set('userId', 'demo-user-id');
    c.set('conversationId', `conv_demo-user-id_${Date.now()}`);

    return await next();
  } catch (error) {
    console.error('Authentication error:', error);
    return c.json({ error: 'Invalid or expired token' }, 401);
  }
};